src/main/java/org/embulk/input/marketo/delegate/ProgramInputPlugin.java

```diff
@@ -0,0 +1,224 @@
+package org.embulk.input.marketo.delegate;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import com.google.common.base.Optional;
+import com.google.common.collect.FluentIterable;
+
+import org.embulk.base.restclient.ServiceResponseMapper;
+import org.embulk.base.restclient.jackson.JacksonServiceResponseMapper;
+import org.embulk.base.restclient.record.ServiceRecord;
+import org.embulk.base.restclient.record.ValueLocator;
+import org.embulk.config.Config;
+import org.embulk.config.ConfigDefault;
+import org.embulk.config.ConfigDiff;
+import org.embulk.config.ConfigException;
+import org.embulk.config.TaskReport;
+import org.embulk.input.marketo.MarketoService;
+import org.embulk.input.marketo.MarketoUtils;
+import org.embulk.spi.Schema;
+import org.embulk.spi.type.Types;
+import org.joda.time.DateTime;
+import org.joda.time.Duration;
+import org.joda.time.format.DateTimeFormat;
+import org.joda.time.format.DateTimeFormatter;
+
+import java.util.Date;
+import java.util.Iterator;
+import java.util.List;
+
+public class ProgramInputPlugin extends MarketoBaseInputPluginDelegate<ProgramInputPlugin.PluginTask>
+{
+    private static final DateTimeFormatter DATE_FORMATER = DateTimeFormat.forPattern(MarketoUtils.MARKETO_DATE_SIMPLE_DATE_FORMAT);
+
+    public interface PluginTask extends MarketoBaseInputPluginDelegate.PluginTask
+    {
+        @Config("query_by")
+        @ConfigDefault("null")
+        Optional<QueryBy> getQueryBy();
+
+        @Config("tag_type")
+        @ConfigDefault("null")
+        Optional<String> getTagType();
+
+        @Config("tag_value")
+        @ConfigDefault("null")
+        Optional<String> getTagVallue();
+
+        @Config("earliest_updated_at")
+        @ConfigDefault("null")
+        Optional<Date> getEarliestUpdatedAt();
+
+        @Config("latest_updated_at")
+        @ConfigDefault("null")
+        Optional<Date> getLatestUpdatedAt();
+
+        @Config("filter_type")
+        @ConfigDefault("null")
+        Optional<String> getFilterType();
+
+        @Config("filter_values")
+        @ConfigDefault("null")
+        Optional<List<String>> getFilterValues();
+
+        @Config("report_duration")
+        @ConfigDefault("null")
+        Optional<Long> getReportDuration();
+
+        void setLatestUpdatedAt(Optional<Date> latestUpdatedAt);
+    }
+
+    public ProgramInputPlugin()
+    {
+    }
+
+    @Override
+    public void validateInputTask(PluginTask task)
+    {
+        super.validateInputTask(task);
+        // validate if query_by is selected
+        if (task.getQueryBy().isPresent() && task.getQueryBy().get() == QueryBy.TAG_TYPE) {
```

Comment:
Too much `if` branches and it will be hard to read, I'd suggest we group then using switch case, something like:

```
if (task.getQueryBy().isPresent()) {
            switch (task.getQueryBy().get()) {
                case TAG_TYPE:
                    // do with tag type
                    break;
                case DATE_RANGE:
                    // do with date range
            }
        }
```